src/tools/kdb/get.c

```diff
@@ -0,0 +1,225 @@
+		keyDel (toLookUp);
+		keyDel (parentKey);
+		ksDel (searchIn);
+		kdbClose (handle, errorKey);
+		RETURN (5)
+	}
+	keyCopyAllMeta (errorKey, parentKey);
+	kdbClose (handle, errorKey);
+
+	setCallback (toLookUp, warnOnMeta);
+	if (verbose)
+	{
+		CLI_PRINT (CLI_LOG_VERBOSE, "got %ld keys\n", ksGetSize (searchIn));
+		setCallback (toLookUp, printTrace);
+	}
+
+	Key * found = ksLookup (searchIn, toLookUp, KDB_O_CALLBACK);
+
+	if (keyCopyAllMeta (errorKey, toLookUp) == -1)
+	{ // we have to copy the meta keys so the CLI can print the errors
+		ELEKTRA_SET_INTERNAL_ERROR (errorKey, "could not copy key meta to errorKey");
+		keyDel (toLookUp);
+		ksDel (searchIn);
+		RETURN (3)
+	}
+
+	if (found == NULL)
+	{
+		CLI_ERROR_PRINT (CLI_LOG_NONE, "Did not find key '%s'", RED (keyName (toLookUp)));
+		ret = 11;
+		goto cleanup;
+	}
+
+	if (keyGetNamespace (found) == KEY_NS_DEFAULT)
+	{
+		CLI_PRINT (CLI_LOG_VERBOSE, "The key was not found in any other namespace, taking the %s\n", BOLD ("default"));
+	}
+	CLI_PRINT (CLI_LOG_VERBOSE, "The resulting keyname is %s\n", keyName (found));
+	CLI_PRINT (CLI_LOG_VERBOSE, "The resulting value size is %ld\n", keyGetValueSize (found));
+
+	if (keyIsBinary (found))
+	{ // for binary data in keys
+		ssize_t binSize = keyGetValueSize (found);
+		CLI_PRINT (CLI_LOG_VERBOSE, "The key is %s.\n", BOLD (binSize == 0 ? "null" : "binary"));
+		const uint8_t * data = (const uint8_t *) keyValue (found);
+		for (int position = 0; position < keyGetValueSize (found); position++)
+		{
+			CLI_PRINT (CLI_LOG_NONE, "\\x%02x", data[position]);
+		}
+	}
+	else
+	{
+		CLI_PRINT (CLI_LOG_NONE, "%s", keyString (found));
+	}
+
+cleanup:
+	if (!noNewLine)
+	{
+		printf ("\n");
+	}
+	keyDel (parentKey);
+	keyDel (toLookUp);
+	ksDel (searchIn);
+
+	RETURN (ret)
+}
+
+void printOptions (elektraLookupFlags options)
+{
+	if (options & KDB_O_SPEC) printf ("KDB_O_SPEC ");
+	if (options & KDB_O_CREATE) printf ("KDB_O_CREATE ");
+	if (options & KDB_O_NOCASCADING) printf ("KDB_O_NOCASCADING ");
+	if (options & KDB_O_NOSPEC) printf ("KDB_O_NOSPEC ");
+	if (options & KDB_O_NODEFAULT) printf ("KDB_O_NODEFAULT ");
+	if (options & KDB_O_CALLBACK) printf ("KDB_O_CALLBACK");
+}
+
+const char * getCascadingName (const char * str)
+{
+	if (str == NULL) return "/";
+	const char * r = strchr (str, '/');
+	return r == NULL ? "/" : r;
+}
+
+Key * warnOnMeta (ELEKTRA_UNUSED KeySet * ks, ELEKTRA_UNUSED Key * key, Key * found, elektraLookupFlags options)
+{
+	if (found != NULL && strncmp (keyName (found), "spec:/", 6) == 0 && options == KDB_O_CALLBACK)
+	{
+		const Key * meta = keyGetMeta (found, "context");
+		if (meta != NULL)
+		{
+			ELEKTRA_ADD_RESOURCE_WARNINGF (
+				key, "%s is context dependent, shown result might be wrong, -v shows you the trace to the key\n",
+				keyName (found));
+		}
+	}
+	return found;
+}
+
+
+Key * printTrace (ELEKTRA_UNUSED KeySet * ks, Key * key, Key * found, elektraLookupFlags options)
```

Comment:
Why add `KeySet * ks` as an argument, if it's not needed?

Reply:
The signature has to look like this for setCallback